src/include/distributed/multi_transaction.h

```diff
@@ -0,0 +1,164 @@
+/*-------------------------------------------------------------------------
+ *
+ * multi_transaction.h
+ *	  Type and function declarations used in performing transactions across
+ *	  the cluster.
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#ifndef MULTI_TRANSACTION_H
+#define MULTI_TRANSACTION_H
+
+#include "distributed/master_metadata_utility.h"
+
+/* maximum (textual) lengths of hostname and port */
+#define MAX_NODE_LENGTH 255 /* includes 0 byte */
+
+/* forward declare, to avoid forcing libpq on everyone */
+struct pg_conn; /* target of the PGconn typedef */
+struct pg_result; /* target of the PGresult typedef */
+
+/* Enumeration that defines different remote transaction states */
+typedef enum
+{
+	/* no transaction active */
+	MULTI_TRANSACTION_STATE_INVALID = 0,
+
+	/* error occurred on connection */
+	MULTI_TRANSACTION_STATE_FAILED = 1 << 0,
+
+	/* transaction start */
+	MULTI_TRANSACTION_STATE_STARTING = 1 << 1,
+	MULTI_TRANSACTION_STATE_STARTED = 1 << 2,
+
+	/* 2pc prepare */
+	MULTI_TRANSACTION_STATE_PREPARING = 1 << 4,
+	MULTI_TRANSACTION_STATE_PREPARED = 1 << 5,
+
+	/* transaction abort */
+	MULTI_TRANSACTION_STATE_ABORTING = 1 << 6,
+	MULTI_TRANSACTION_STATE_ABORTED = 1 << 7,
+
+	/* transaction commit */
+	MULTI_TRANSACTION_STATE_COMMITTING = 1 << 8,
+	MULTI_TRANSACTION_STATE_COMMITTED = 1 << 9,
+
+	MULTI_TRANSACTION_STATE_DDL = 1 << 10,
+	MULTI_TRANSACTION_STATE_DML = 1 << 11
+} MultiTransactionState;
+
+
+typedef struct MultiConnection
+{
+	char hostname[MAX_NODE_LENGTH];
+	int32 port;
+	char user[NAMEDATALEN];
+	char database[NAMEDATALEN];
+
+	char prepared_name[NAMEDATALEN];
+
+	bool sessionLifespan;
+
+	bool busy;
+
+	bool activeInXact;
+
+	struct pg_conn *conn;
+
+	/* move to private part of struct? */
+	uint32 transactionState;
```

Comment:
On 2016-09-08 05:56:42 -0700, Marco Slot wrote:
> > +	int32 port;
> > +	char user[NAMEDATALEN];
> > +	char database[NAMEDATALEN];
> > +
> > +	char prepared_name[NAMEDATALEN];
> > +
> > +	bool sessionLifespan;
> > +
> > +	bool busy;
> > +
> > +	bool activeInXact;
> > +
> > +	struct pg_conn *conn;
> > +
> > +	/* move to private part of struct? */
> > +	uint32 transactionState;
> 
> Could we use a single enum value for the transactionState (augmented with flags if necessary)? I find the bit flipping a bit harder to follow.

I had that originally, and I found it a lot harder to read, and it's
harder to debug, because "history" is lost. So I'd prefer to not go
there.